eos-update-flatpak-repos

```diff
@@ -624,74 +548,71 @@ def _migrate_installed_flatpaks():
+            assert (old_ostree_ref != new_ostree_ref)
 
             try:
-                new_branch = migrate.get('new-branch')
-                if new_branch:
+                # copy the old ref to the new one, pointing at the same commit
+                if old_ostree_ref in ostree_refs and \
+                   not new_ostree_ref in ostree_refs:
+                    logging.info('Copying OSTree ref {} [{}] to {}'
+                                 .format(old_ostree_ref,
+                                         ostree_refs[old_ostree_ref],
+                                         new_ostree_ref))
+                    repo.set_ref_immediate(new_origin,
+                                           new_ostree_ref.split(':')[1],
+                                           ostree_refs[old_ostree_ref])
+                    ostree_refs[new_ostree_ref] = ostree_refs[old_ostree_ref]
+
+                if new_branch != branch:
                     try:
-                        new_ref = inst.get_installed_ref(kind, name, arch,
-                                                         new_branch)
+                        ref = inst.get_installed_ref(kind, name, arch,
+                                                     new_branch)
+                        logging.info('Found already installed: {}'
+                                     .format(ref.format_ref()))
+
+                        # adopt installed origin if app was already installed
+                        new_origin = ref.get_origin()
                     except GLib.GError as e:
                         if not e.matches(Flatpak.error_quark(),
                                          Flatpak.Error.NOT_INSTALLED):
                             raise
-                    else:
-                        logging.info('{} already installed, removing {}'
-                                     .format(new_ref.format_ref(),
-                                             ref.format_ref()))
-                        inst.uninstall_full(Flatpak.UninstallFlags.NO_PRUNE,
-                                            kind, name, arch, branch)
-
-                        # do not continue to do any origin migrations if we already
-                        # have the app installed. the ostree ref is already part
-                        # of matching_ostree_refs so will be removed at the end.
-                        continue
-
-                    _update_branch(ref, new_branch, refs)
-                    branch = new_branch
-                    ref = _refresh_ref(inst, kind, name, arch, branch)
 
-                deploy_dir = ref.get_deploy_dir()
-                deploy = os.path.join(deploy_dir, 'deploy')
+                        logging.info('Deploying with new ref: {}'.format(new_ostree_ref))
+                        ref = inst.install_full(Flatpak.InstallFlags.NO_PULL, new_origin,
+                                                kind, name, arch, new_branch)
```

Comment:
So... Flatpak checks xa.ref when deploying, and emits a WARNING if this has the wrong branch ID. This warning is promoted to a fatal error if collection IDs are set, which is sadly a merge blocker for this branch.

Reply:
Fixed using code from ostree-upload-server. Thanks @dbnicholson !